File: network/src/main/java/com/ayvytr/network/CacheInterceptor.java
package com.ayvytr.network;

import android.content.Context;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * @author ayvytr
 */
public class CacheInterceptor implements Interceptor {

    private Context context;
    private int maxStaleSeconds;

    public CacheInterceptor(Context context) {
        this(context, 7 * 24 * 60 * 60);
    }

    public CacheInterceptor(Context context, int maxStaleSeconds) {
        this.context = context;
        this.maxStaleSeconds = maxStaleSeconds;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        if(NetworkUtils.isConnected(context)) {
            return chain.proceed(chain.request());
        } else { // 如果没有网络，则返回缓存数据
            Request newRequest = chain.request().newBuilder()
                                      .removeHeader("Pragma")
                                      .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStaleSeconds)
                                      .build();
            return chain.proceed(newRequest);
        }
    }
}

File: network/src/main/java/com/ayvytr/network/interceptor/CacheInterceptor.kt
package com.ayvytr.network.interceptor

import com.ayvytr.network.isNetworkAvailable
import okhttp3.CacheControl
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.util.concurrent.TimeUnit

internal class CacheInterceptor(val maxAgeSeconds: Int = 3600) : Interceptor {
    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val response: Response
        val request: Request = if (isNetworkAvailable()) {
            //有网络,检查缓存
            chain.request()
                .newBuilder()
                .cacheControl(CacheControl.Builder()
                                  .maxAge(maxAgeSeconds, TimeUnit.SECONDS)
                                  .build())
                .build()
        } else {
            //无网络,检查缓存,即使是过期的缓存（这里和maxAge一致，有需要再改）
            chain.request().newBuilder()
                .cacheControl(CacheControl.Builder()
                                  .onlyIfCached()
                                  .maxStale(maxAgeSeconds, TimeUnit.SECONDS)
                                  .build())
                .build()
        }

        response = chain.proceed(request)
        return response.newBuilder().build()
    }
}